dataplane/internal/engine/port.go

```diff
@@ -0,0 +1,282 @@
+						ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
+						Action: &fwdpb.ActionDesc_Lookup{
+							Lookup: &fwdpb.LookupActionDesc{
+								TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: puntEtherTypeTable(name)}},
+							},
+						},
+					}, { // Check IP Proto punt rules.
+						ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
+						Action: &fwdpb.ActionDesc_Lookup{
+							Lookup: &fwdpb.LookupActionDesc{
+								TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: puntIPProtoType(name)}},
+							},
+						},
+					}, { // Lookup in FIB.
+						ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
+						Action: &fwdpb.ActionDesc_Lookup{
+							Lookup: &fwdpb.LookupActionDesc{
+								TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: fibSelectorTable}},
+							},
+						},
+					}},
+					Outputs: []*fwdpb.ActionDesc{{ // Lookup in the to port to src MAC table.
+						ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
+						Action: &fwdpb.ActionDesc_Lookup{
+							Lookup: &fwdpb.LookupActionDesc{
+								TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: srcMACTable}},
+							},
+						},
+					}},
+				},
+			},
+		},
+	}
+	if _, err := c.PortUpdate(ctx, update); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateLocalPort creates an local (ie TAP) port.
+func CreateLocalPort(ctx context.Context, c fwdpb.ServiceClient, name string) error {
+	port := &fwdpb.PortCreateRequest{
+		ContextId: &fwdpb.ContextId{Id: contextID},
+		Port: &fwdpb.PortDesc{
+			PortType: fwdpb.PortType_PORT_TYPE_KERNEL,
+			Port: &fwdpb.PortDesc_Kernel{
+				Kernel: &fwdpb.KernelPortDesc{
+					DeviceName: name,
+				},
+			},
+		},
+	}
+	if _, err := c.PortCreate(ctx, port); err != nil {
+		return err
+	}
+
+	update := &fwdpb.PortUpdateRequest{
+		ContextId: &fwdpb.ContextId{Id: contextID},
+		PortId:    &fwdpb.PortId{ObjectId: &fwdpb.ObjectId{Id: name}},
+		Update: &fwdpb.PortUpdateDesc{
+			Port: &fwdpb.PortUpdateDesc_Kernel{
+				Kernel: &fwdpb.KernelPortUpdateDesc{
+					Inputs: []*fwdpb.ActionDesc{{ // Lookup in FIB.
+						ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
+						Action: &fwdpb.ActionDesc_Lookup{
+							Lookup: &fwdpb.LookupActionDesc{
+								TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: fibSelectorTable}},
+							},
+						},
+					}},
+				},
+			},
+		},
+	}
+	if _, err := c.PortUpdate(ctx, update); err != nil {
+		return err
+	}
+	return nil
+}
+
+// puntEtherTypeTable returns the name of the table containing the punt rules based on the EtherType packet header field.
+func puntEtherTypeTable(port string) string {
+	return fmt.Sprintf("%s-punt-etherType", port)
+}
+
+// puntEtherTypeTable returns the name of the table containing the punt rules based on the IP protocol packet header field.
+func puntIPProtoType(port string) string {
+	return fmt.Sprintf("%s-punt-ipproto", port)
+}
+
+func setupPuntRules(ctx context.Context, c fwdpb.ServiceClient, name string) error {
```

Comment:
setupPuntTables to use the same "table" word as SetupTables? Maybe rename that one to SetupForwardingTables?